main.cpp

```diff
@@ -38,8 +39,275 @@
+	    "h6lqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq"
+	    "qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqxnht6w";
+
+    size_t output_size = std::min(max_size, dummy.size());
+    std::memcpy(fuzz_data, dummy.data(), output_size);
+    return output_size;
+}
+
+// We use a custom mutator to produce an input corpus that consists entirely of
+// correctly encoded bech32 strings. This enables us to efficiently fuzz the
+// bolt11 decoding logic without the fuzzer getting stuck on fuzzing the bech32
+// decoding/encoding logic.
+//
+// This custom mutator does the following things:
+//   1. Attempt to bech32 decode the given input (returns the encoded dummy
+//      invoice on failure).
+//   2. Mutate either the human readable or data part of the invoice using
+//      libFuzzer's default mutator `LLVMFuzzerMutate`.
+//   3. Attempt to bech32 encode the mutated hrp and data (returns the endcoded
+//      dummy on failure).
+//   4. Write the encoded result to `fuzz_data` if its size does not exceed
+//      `max_size`, otherwise return the encoded dummy invoice.
+extern "C" size_t LLVMFuzzerMutate(uint8_t *Data, size_t Size, size_t MaxSize);
+
+size_t LLVMFuzzerCustomMutator(uint8_t *fuzz_data, size_t size, size_t max_size,
+                               unsigned int seed)
+{
+    // A minimum size of 9 prevents hrp_maxlen <= 0 and data_maxlen <= 0.
+    if (size < 9)
+        return initial_input(fuzz_data, size, max_size);
+
+    // Interpret fuzz input as string
+    std::string input(reinterpret_cast<char*>(fuzz_data), size);
+
+    // Attempt to bech32 decode the input
+    bech32::DecodeResult decoded = bech32::Decode(input, bech32::CharLimit::CUSTOM_MUTATOR);
+    if (decoded.encoding != bech32::Encoding::BECH32) {
+        // Decoding failed, this should only happen when starting from
+        // an empty corpus.
+        return initial_input(fuzz_data, size, max_size);
+    }
+
+    auto data = decoded.data;
+    auto hrp = decoded.hrp;
+
+    size_t data_maxlen = input.size() > 8 ? std::min(input.size() - 8, static_cast<size_t>(65)) : 0;
+    size_t hrp_maxlen = input.size() > 6 ? std::min(input.size() - 6, static_cast<size_t>(83)) : 0;
+
+    // Mutate either the hrp or data
+    std::srand(seed);
+    switch (std::rand() % 2) {
+        case 0: { // Mutate hrp
+            if (hrp_maxlen > 0) {
+                // Make sure we have a buffer that's large enough for mutation
+                std::vector<uint8_t> hrp_buffer(hrp.begin(), hrp.end());
+                // Reserve enough space for the maximum mutation size
+                hrp_buffer.resize(hrp_maxlen);
+
+                // Mutate the buffer
+                size_t new_len = LLVMFuzzerMutate(hrp_buffer.data(),
+                                                 hrp.size(),
+                                                 hrp_maxlen - 1);
+
+                // Ensure minimum length and validate - at least 1 character
```

Comment:
If libFuzzer truncates the HRP or data part to length 0, why do we ignore that and enforce len 1 instead?